File: utbot-spring-analyzer/src/main/java/utils/ConfigurationManager.java
package utils;

import org.springframework.context.annotation.ImportResource;
import org.springframework.context.annotation.PropertySource;

import java.lang.annotation.Annotation;
import java.lang.reflect.Field;
import java.lang.reflect.InvocationHandler;
import java.util.Arrays;
import java.util.Map;
import java.util.Optional;

public class ConfigurationManager {

    private final ClassLoader classLoader;
    private final Class userConfigurationClass;

    public ConfigurationManager(ClassLoader classLoader, Class userConfigurationClass) {
        this.classLoader = classLoader;
        this.userConfigurationClass = userConfigurationClass;
    }

    public void patchPropertySourceAnnotation() throws Exception {
        patchAnnotation(PropertySource.class, String.format("classpath:%s", ResourceNames.fakePropertiesFileName));
    }

    public void patchImportResourceAnnotation() throws Exception {
        patchAnnotation(ImportResource.class, String.format("classpath:%s", ResourceNames.fakeApplicationXmlFileName));
    }

    private void patchAnnotation(Class<?> annotationType, String newValue) throws Exception {
        Class<?> proxyClass = classLoader.loadClass("java.lang.reflect.Proxy");
        Field hField = proxyClass.getDeclaredField("h");
        hField.setAccessible(true);

        Optional<Annotation> propertySourceAnnotation =
                Arrays.stream(userConfigurationClass.getAnnotations())
                        .filter(el -> el.annotationType() == annotationType)
                        .findFirst();

        if (propertySourceAnnotation.isPresent()) {
            InvocationHandler annotationInvocationHandler = (InvocationHandler) (hField.get(propertySourceAnnotation.get()));

            Class<?> annotationInvocationHandlerClass =
                    classLoader.loadClass("sun.reflect.annotation.AnnotationInvocationHandler");
            Field memberValuesField = annotationInvocationHandlerClass.getDeclaredField("memberValues");
            memberValuesField.setAccessible(true);

            Map<String, Object> memberValues = (Map<String, Object>) (memberValuesField.get(annotationInvocationHandler));
            memberValues.put("value", newValue);
        }
    }
}

File: utbot-spring-analyzer/src/main/kotlin/utils/ConfigurationManager.kt
package utils

import org.springframework.context.annotation.ImportResource
import org.springframework.context.annotation.PropertySource
import java.lang.reflect.InvocationHandler
import java.nio.file.Path
import java.util.Arrays
import kotlin.reflect.KClass

class ConfigurationManager(private val classLoader: ClassLoader, private val userConfigurationClass: Class<*>) {

    fun clearPropertySourceAnnotation(){
        patchAnnotation(
            PropertySource::class, null
        )
    }

    fun clearImportResourceAnnotation(){
        patchAnnotation(
            ImportResource::class, null
        )
    }

    fun patchPropertySourceAnnotation(userPropertiesFileName: Path) =
        patchAnnotation(
            PropertySource::class, String.format(
                "classpath:%s", "fake_$userPropertiesFileName"
            )
        )

    fun patchImportResourceAnnotation(userXmlFilePath: Path) =
        patchAnnotation(
            ImportResource::class,
            String.format(
                "classpath:%s", "fake_$userXmlFilePath"
            )
        )

    private fun patchAnnotation(annotationClass: KClass<*>, newValue: String?) {
        val proxyClass = classLoader.loadClass("java.lang.reflect.Proxy")
        val hField = proxyClass.getDeclaredField("h")
        hField.isAccessible = true

        val propertySourceAnnotation = Arrays.stream(
            userConfigurationClass.annotations
        )
            .filter { el: Annotation -> el.annotationClass == annotationClass }
            .findFirst()

        if (propertySourceAnnotation.isPresent) {
            val annotationInvocationHandler = hField[propertySourceAnnotation.get()] as InvocationHandler

            val annotationInvocationHandlerClass =
                classLoader.loadClass("sun.reflect.annotation.AnnotationInvocationHandler")
            val memberValuesField = annotationInvocationHandlerClass.getDeclaredField("memberValues")
            memberValuesField.isAccessible = true

            val memberValues = memberValuesField[annotationInvocationHandler] as MutableMap<String, Any>
            addNewValue(memberValues, newValue)
        }
    }

    private fun addNewValue(memberValues: MutableMap<String, Any>, newValue: String?){
        if(newValue == null){
            memberValues["value"] = Array(0){""}
        }
        else {
            val list: MutableList<String> = (memberValues["value"] as Array<String>).toMutableList()
            list.add(newValue)
            memberValues["value"] = list.toTypedArray()
        }
    }
}